File: my_dart/js/cascade.js
// NOTE：這邊是js的語法
class DateTime {
    _date = null;

    constructor() {
        this._date = new Date();
    }

    add(num, type) {
        switch (type) {
            case 'days':
                this._addDays(num);
                break;
            default:
        }

        return this;
    }

    setDate(newDate) {
        _date = newDate;
        return _date;
    }

    _addDays(num) {
        this._date.setDate(this._date.getDate() + num);
    }

    toString() {
        return this._date.toString();
    }
}

const date = new DateTime();

// 函式鏈結
const result =
    date
    .setDate(new Date(2021, 8, 2, 0, 0, 0))
    .add(30, 'days')
    .toString();

console.log(result); // Sat Oct 02 2021 00:00:00 GMT+0800 (台北標準時間)

File: my_dart/cascade.dart
class MyDateTime {
  DateTime _date;

  MyDateTime() : _date = DateTime.now();

  add(int num, String type) {
    switch (type) {
      case 'days':
        _addDays(num);
        break;
      default:
    }
  }

  setDate(DateTime date) {
    this._date = date;
    return this._date; // 使用(.)與(..)取得的內容會不一樣
  }

  _addDays(int num) {
    _date = _date.add(Duration(days: num));
  }

  toString() {
    return _date.toString();
  }
}

void main() {
  var date = MyDateTime();

  var result = date
    ..setDate(DateTime(2021, 8, 2))
    ..add(30, 'days')
    ..toString();

  print(result); // 2021-09-01 00:00:00.000

  var dateTime = MyDateTime().setDate(DateTime(2021, 8, 2));
  print(dateTime.runtimeType); // DateTime
  var myDateTime = MyDateTime()..setDate(DateTime(2021, 8, 2));
  print(myDateTime.runtimeType); // MyDateTime
}
